### ISSUE TITLE
install of unwanted version: 1.1 & 1.10

### ISSUE THREAD

**Issue by flav1972** [tfutils/tfenv#432]
installing latest:^1.1
installed in fact 1.10.0-alpha20240606

**Comment by MageshSrinivasulu** [comment]
This issue needs an immediate fix

**Comment by flav1972** [comment]
```
$ cat .terraform-version 
$ latest:^1.1
$ tfenv install
Installing Terraform v1.10.0-alpha20240606
```
and

```
$ cat .terraform-version 
latest:^1\.1\.
$ tfenv install
Installing Terraform v1.1.9
```

At least documentation should be updated.

**Comment by Zordrak** [comment]
I agree the documentation examples should make it clearer that this is a string match and that `^0.8` would match `0.8.1` but also `0.81.0`.

Will do.

**Comment by Zordrak** [comment]
## Root Cause Analysis (from code review)

The issue is in the regex construction for `latest:^X.Y` patterns.

When you request `latest:^1.1`, the regex `^1.1` is used as a grep pattern against the version list. In regex, `.` matches ANY character, so `^1.1` matches:
- `1.1.0` through `1.1.9` (intended)
- `1.10.0` through `1.10.x` (unintended - the `.` matches the `0` in `10`)
- Even things like `1210` if they existed

The fix is to escape the dots in the version number before using it as a regex, or use `^1\.1\.` (with escaped dots and a trailing dot) to ensure it only matches the `1.1.x` series:

```bash
# Instead of:
regex="^1.1"
# Use:
regex="^1\.1\."
```

This affects all `latest:^X.Y` patterns where Y is a prefix of a larger number (1.1 vs 1.10, 1.2 vs 1.20, etc).

**Comment by Zordrak** [comment]
Closing — this is working as designed. When you use `latest:^1.1` in your `.terraform-version` file, the `^1.1` part is a *regular expression*, not a version prefix. In regex, `.` matches any character, so `^1.1` matches `1.10.x`, `1.11.x`, etc. To match only 1.1.x versions, escape the dot: `latest:^1\.1\.`. This is documented in the README under the `.terraform-version` file section.
